docs(concepts): add under the hood (#3581)
webpack/webpack.js.org: src/content/concepts/under-the-hood.md

---
title: Under The Hood
sort: 1
contributors:
  - smelukov
---

> This section describes webpack internals and useful for plugin developers

> The bundling is a function that takes some files and emits others.

Reviewer: What do you mean?

Author: It's a very basic definition of bundling

Reviewer: ```suggestion
Bundling is a function that takes some files and emits others.
```

But between input and output, it also has [modules](/concepts/modules/), [entry points](/concepts/entry-points/), chunks, chunk groups, and many other intermediate parts.

## The main parts

Every file used in your project is a [Module](/concepts/modules/)

__./index.js__

```js
import app from './app.js';
```

__./app.js__

```js
export default 'the app';
```

> By using each other, the modules form a graph (`ModuleGraph`). 

Reviewer: Is there any docs to explain what a ModuleGraph is?

Author: Not yet. In progress

> Usually, modules don't exist by themselves and combine into chunks.

Reviewer: No sure what you mean here.

Author: It fixed by #3586

Each of them has a chunk with a module - `./home.js` for `home` and `./about.js` for `about`

> There are might be more than one chunk in the chunk group. For example [SplitChunkPlugin](/plugins/split-chunks-plugin/) do so - it splits a chunk into one or more chunks.

## Chunks

Chunks come in two forms - `initial` and `non-initial`

`initial` is the main chunk for the entry point. This chunk contains all the modules (and its dependencies) that you specify for an entry point.

`non-initlal` is a chunk that may be lazy-loaded. It may appear when [dynamic import](/guides/code-splitting/#dynamic-imports) or [SplitChunkPlugin](/plugins/split-chunks-plugin/) is using.

Each chunk has a corresponding __asset__.
The assets are the output files - the result of the bundling.

__webpack.config.js__

```js
module.exports = {
  entry: './src/index.jsx'
};
```

__./src/index.js__

```js
import React from 'react';
import ReactDOM from 'react-dom';

import('./app.jsx').then(App => ReactDOM.render(<App />, root));
```

Initial chunk with name `main` created. It contains:

- `./src/index.jsx`
- `react`
- `react-dom`

and all their dependencies, except `./app.jsx`

> Non-initial chunk for `./app.jsx` created because this module imported dynamically.

Reviewer 2: ```suggestion
Non-initial chunk for `./app.jsx` is created as this module is imported dynamically.
```

__Output:__

- `/dist/main.js` - an initial chunk 
- `/dist/394.js` - non-initial chunk

> By default, there is no name for non-initial chunks so that a unique ID used instead of a name.

Reviewer 2: ```suggestion
By default, there is no name for non-initial chunks so that a unique ID is used instead of a name.
```
